### DISCUSSION TITLE
Differences between running nextjs out folder (export) with npx serve and PM2 serve

### DISCUSSION THREAD

**Discussion by egachi** [vercel/next.js#34934] [Help]
When serving just static pages using `next build && next export`, there is a big difference between `npx serve` and `pm2 serve`. 

Having the following configuration: 

-  `next.config.js`

```javascript
/**
 * @type {import('next').NextConfig}
 */

 const nextConfig = {
     
    trailingSlash: true
  }
  
  module.exports = nextConfig
```

-  `package.json`

```javascript

{
  "name": "helloworld",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "export": "next export",
    "serve": "npx serve out",
    "pm2": "pm2 serve out --no-daemon",
    "http": "http-server out"
  },
  "dependencies": {
    "next": "12.1.0",
    "react": "17.0.2",
    "react-dom": "17.0.2"
  },
  "devDependencies": {
    "eslint": "8.9.0",
    "eslint-config-next": "12.1.0"
  }
}

```

- Using `npx serve out`, testing these approaches:
  - Doing request directly to `http://localhost:3000/admin` *(working)*
  - Doing request directly to `http://localhost:3000/admin/` *(working)*
  - Or clicking from home and use a `next/link` to admin, it is working. *(working)*

- Using `pm2 serve out --no-daemon`, testing these approaches:
  - Doing request directly to `http://localhost:3000/admin` *(not working - 404 error)*
  - Doing request directly to `http://localhost:3000/admin/` *(not working - 404 error)*
  - Or clicking from home and use a `next/link` to admin, it is working. *(working)*
  - With using `pm2 serve out --no-daemon --spa` is redirecting all requests to index.html *(expected)* 

Is there any consideration or workaround for pm2 on this scenario or do we need to discard the pm2 option to serve static pages for nextjs framework?

**Comment by ghost** [comment]
It took me hours to figure out the workarround for next js problem with PM2. 
Thanks @edisga for raising this issue. I think that was the longest debugging session I had with you on next js routing with PM2, if you remember it.  I hope nextjs team will work on it.

**Reply by navalpatel5** [reply]
Wow... i am elated to find this article as I am in the same boat. But could not find the solution. If there is any please let me know.

**Comment by SouthLink** [comment]
You can try skipTrailingSlashRedirect this configuration, it can make your service support http://localhost:3000/admin and http://localhost:3000/admin/ both modes：

```
/**
 * @type {import('next').NextConfig}
 */

 const nextConfig = {
    trailingSlash: true,
    skipTrailingSlashRedirect: true
  }
  
  module.exports = nextConfig
```

If this doesn't solve your problem, you can try creating pm2.json in the root directory

```
{
  "apps": [
    {
      "name": "you nanme",
      "script": "serve out",
      "instances": 5,
      "instance_var": "INSTANCE_ID",
      "exec_mode": "cluster_mode",
      "env": {
        "NODE_ENV": "production"
      },
      "max_memory_restart": "700M"
    }
  ]
}
```

**pm2 start pm2.json**，do this to start the project
